op-node/p2p/store/records_book.go

```diff
@@ -80,26 +84,24 @@ func (d *recordsBook[K, V]) startGC() {
 	startGc(d.ctx, d.log, d.clock, &d.bgTasks, d.prune)
 }
 
-func (d *recordsBook[K, V]) GetRecord(key K) (V, error) {
-	d.RLock()
-	defer d.RUnlock()
-	rec, err := d.getRecord(key)
-	return rec, err
-}
-
 func (d *recordsBook[K, V]) dsKey(key K) ds.Key {
 	return d.dsBaseKey.Child(d.dsEntryKey(key))
 }
 
 func (d *recordsBook[K, V]) deleteRecord(key K) error {
-	d.cache.Remove(key)
+	// If access to this isn't synchronized, removing from the cache first can result in the stored
+	// item being cached again before it is deleted.
 	err := d.store.Delete(d.ctx, d.dsKey(key))
+	d.cache.Remove(key)
 	if err == nil || errors.Is(err, ds.ErrNotFound) {
 		return nil
 	}
 	return fmt.Errorf("failed to delete entry with key %v: %w", key, err)
 }
 
+// You must read lock the recordsBook before calling this, and only unlock when you have extracted
+// the values you want from the value of type V. There's no way to conveniently pass an extractor
+// function parameterized on V here without breaking this out into a top-level function.
 func (d *recordsBook[K, V]) getRecord(key K) (v V, err error) {
 	if val, ok := d.cache.Get(key); ok {
```

Comment:
Is there a reason as to why the lock cannot be grabbed internal to this function? That would remove the footgun around callers forgetting

Reply:
Yeah see above, the reason is a pointer to the value is held after the function returns and then values extracted from it. The recordsBook is an internal type to the package anyway so it's harmless and clearer to allow callers to specify the kind of locking they need and when. (For example SetRecord also needs to lock with write permission, and then calls getRecord so that's a bad day if it locks internally).